### DISCUSSION TITLE
Getting old profile when retrieving different users

### DISCUSSION THREAD

**Discussion by rfossella** [node-saml/node-saml#382] [Q&A] [answered]
Hi,

Using  
"passport": "^0.5.3",
"passport-saml": "^3.2.4",
"express": "^4.18.2",
"express-session": "^1.17.3",
node 16.20

Browsers 
Chrome Version 129.0.6668.90 (Official Build) (64-bit)
Edge Version 129.0.2792.89 (Official build) (64-bit)

**I THOUGHT THIS ISSUE WAS ONLY IN CHROME, BUT IT OCCURS IN EDGE TOO**

---
I have a passport-saml set up whereby I have to support **multiple** sets of credentials - with IdP MS Entra ID

Because of this, in my /login route I dynamically get the creds and create new SamlStrategy

```
router.post('/login', (req, res, next) => 
{
    const username = req.body.username;
    const role= req.body.role; 
    const action = req.body.action;

passport.use(new SamlStrategy(getSAMLCredentials(username), 
		(profile, done) =>
		{
			console.log('SAML Profile:', profile);
		}))

// Initiate SAML login
passport.authenticate('saml')(req, res, next);

```
I have three (3) users.  When I **first** login USER1 I properly get prompted through to MS Entra to select user name, then password.
I then correctly hit callback code with a valid profile.   I then correctly hit my login/callback route via MS Entra.  I do my req.login(user..) and user is successfully logged into my application.

I then logout that user using:
```
req.logOut();
req.session.destroy(err => 
{
if (err) {
		console.error('Failed to destroy session:', err);
} else {
		console.log('Session successfully destroyed');
}
	});
	
res.clearCookie('connect.sid', { path: '/' });
```
User is logged out and session shows to be destroyed.

I go back to my login screen.

When I then enter USER2 it appears that I don't get back out to MS Entra to select THAT user name and instead I am still holding onto the OLD/LAST profile from USER1.  I have extra logic to make sure that _entered_ user name matches the _profile_ user name  - this certainly fails because again I am still holding USER1 profile.

If I manually logout of MS Entra - either from browser or programmatically, and then enter USER2, I get a new (correct) prompt for Entra and new fresh profile.  But the cycle continues w/USER3, etc.

As a work-around, I added the following:

```
passport.use(new SamlStrategy({
				...getSAMLCredentials(username),
				forceAuthn: true
			  }, (profile, done) =>
					{
						console.log('SAML Profile:', profile); 
                                     })
```

This is now FORCING a SAML (SSO) request every single time (correctly, based on documentation) but this defeats the purpose of a Single-Sign-On implementation where you only want to have the user logged in once - unless you explicitly log out of IdP.

**Expected behavior**
So, my expectation is:

1. If I FIRST-TIME login USER1, I get the whole Entra user/password check and user gets into application.
2. I then log USER1 out (of application, not Entra).  
3. If I login USER1 again he should go directly into application as he SSO'd already.
4. If I then FIRST-TIME login USER2, I should be again directed to Entra user/pwd, get a new profile back - USER2 gets logged in
5. I log USER2 out of application only
6. Back to USER1 - should go right in... etc.

So again, I am stuck on the thing where I seem to holding onto an old profile even though I am destroying the session , req.logout(), etc.

Is there some else I should be doing w/respect to passport-saml to get clean profiles?  Is this something strange going on w/IdP?  Browsers and cache?

**Comment by srd90** [comment]
First of all you are using deprecated `passport-saml` package so chances that anyone would introduce any fix to that are near zero even if it would turn out that [this @node-saml/node-saml package bug report](https://github.com/node-saml/node-saml/issues/379) would indicate any actual issue.

Based on your story at issue report you you/your code did not trigger SLO when you logged out USER1 so user is still logged in at IdP. If user is still logged in at IdP it works as you described and situation is normal from passport-saml point of view.

Use some SAML tracer plugin at your browser to see whats being transferred between IdP and SP. E.g. this https://chromewebstore.google.com/detail/saml-tracer/mpdajninpobndbfcldcmbpnnbhibjmch

IMHO [this](https://github.com/node-saml/node-saml/issues/379) should have been discussion instead of bug report.

BTW. about achieving fully functional SLO (also reacting correctly to situation when IdP propagates SLO triggered by some other SP to your SP) see: https://github.com/node-saml/passport-saml/blob/v5.0.0/README.md#slo-single-logout

**Comment by rfossella** [comment]
Thanks for your valuable feedback. I agree - this should have been a discussion. Maybe I should close this and create discussion thread?

Interesting about the deprecation.  I actually went to the Passport home page (https://www.passportjs.org/packages/passport-saml/) and followed their current instructions and ended up w/v3.2.4.  I will certainly install @node-saml/passport-saml

Regarding the SLO - you are correct.  When I programmatically trigger the IdP SLO (along with application logout) it will work fine.  However, when the user logs back in they have to re-login to IdP, which is something we wanted to avoid. 

I will read through the SLO link you gave and also the use of MultiSamlStrategy which may help.

Appreciate  your feedback!

**Comment by srd90** [comment]
>Interesting about the deprecation. I actually went to the Passport home page (https://www.passportjs.org/packages/passport-saml/) and followed their current instructions and ended up w/v3.2.4.

One of the [node-saml](https://github.com/node-saml/) organization maintainers has requested to update passportjs catalog information back in Nov 2022: https://github.com/jaredhanson/passport/issues/916#issuecomment-1312595503

Unscoped passport-saml was marked deprecated when 4.0.0 was released.

If I remember correctly npm reports deprecation when installing [package that has been marked deprecated](https://www.npmjs.com/package/passport-saml) at repository and some vulnerability scanners (if such tools are used) might also report deprecated dependencies along with vulnerable dependencies. I.e. it should not be surprise that unscoped passport-saml is/was deprecated.

**Comment by rfossella** [comment]
An issue arose when using the new version whereby it complains about **idpCert not found** - even though using same syntax - i.e.

```
let samlCredentials = getSAMLCredentials(username);
passport.use(new SamlStrategy({...samlCredentials, forceAuthn: false}, 
(profile, done) =>
{
console.log('SAML Profile:', profile); 
})
```

Works fine with **passport-saml**;  fails with **@node-saml/passport-saml**, even though the doc on new page shows same syntax.  My certs are exactly the same.

I read something on stack about needing, wantAssertionsSigned:  and wantAuthnResponseSigned: but had no effect.

I 
```
return {
		entryPoint,
		issuer,
		cert,
		callbackUrl,
		logoutUrl,
		wantAssertionsSigned: false,
		wantAuthnResponseSigned: false
		}
```

Trying the find a syntactical difference in docs as to why this would fail w/new package.

**Comment by srd90** [comment]
>An issue arose when using the new version whereby it complains about idpCert not found - even though using same syntax...

Because after the passport-saml split to @node-saml/node-saml and @node-saml/passport-saml (latter uses former internally) there are still some corners to be updated at documentation. About constructor parameters: readme clearly points to right direction: https://github.com/node-saml/passport-saml/blob/v5.0.0/README.md#configure-strategy

IIRC it was release 5.0.0 that changed cert to idpCert (see release notes of @node-saml/node-saml)

>I read something on stack about needing, wantAssertionsSigned: and wantAuthnResponseSigned: but had no effect.

You tried to resolve "idpCert not found" with those config parameters? I.e. copy pasted this and that to security related component configuration without taking time to understand implications?

FWIW if you want to read more about issues that can be solved by altering values of those "want..." configuration parameters read through this list https://github.com/node-saml/node-saml/issues/378#issuecomment-2344755012

**Comment by rfossella** [comment]
Firstly, thank you for the reference.  I had read that and focused on the first example, seeing the same syntax using "cert".  I did not make the connection referencing ADFS and using "idpCert."  This has since worked for me.


> You tried to resolve "idpCert not found" with those config parameters? I.e. copy pasted this and that to security related component configuration without taking time to understand implications

The stack comments I read specifically pointed to certificate issues.  Reading through lots of posts.  So yes, I must confess, while it certainly seemed unrelated, sadly I was _grasping_ here to see if that config option needed. 

Thanks again for your help.

**Comment by rfossella** [comment]
I was looking for #378 which appears related to a new error I am seeing "Error: Invalid document signature\n    at SAML.validatePostResponseAsync "

Using my previous use case, this occurs when I initially choose USER1, then follow up by using USER2. The second user fails with that error.

**Comment by srd90** [comment]
The way you describe your use case is very odd. You speak about getCredentials when you get configuration for passport-saml from some config store / build it dynamically based on username.

Furthermore if I interpret your first post correctly you are back in the situation where you do only local logout to USER1 and expect exactly same IdP instance to offer login screen to USER2 while user1 is still logged in to IdP and the way single sign on (SSO) works by definition it - IdP - doesn't assert with login screen if it can satisfy login automatically (from SP point of view it responds with user1's authnresponse). Btw. If your user2 is using different "realm" than user1 or whatever they call it at MS entra those realms might have different certs and when combined with how you seem to be fighting against SSO principles ... 

Unless you provide minimal but fully functional source code of your application which you can use to reproduce your issue internally AND also recorded SAML message flow (sensitive parts redacted) from the very beginning of your use case to the bitter end there is no point to continue this **discussion** because it involves too much guessing.

Few more tips until we **might** meet again **when** aforementioned assets are available.

Because you seem to be recreating SAML config per request there is a chance that you have copy pasted some config parameter differently for other user. If you use same IdP instance it could be incorrect IdP cert or other user might have different "wants" due to not copy pasting config correctly from one user to another...see what sort of guessing game this is without proper input for remote debugging via GH issue.

Which leads to second tip: use debugger to step debug to the core of your exact issue. Pay attentention to the config values of wants, certificates used to validate signatures etc. If you do not know where to introduce your first debugger breakpoint may I recommend this line:
https://github.com/node-saml/node-saml/blob/72a39dcf46a93a8679123db78b22ff82d6dd4096/src/saml.ts#L709
and start debugging your scenario.

**Comment by markstos** [comment]
I opened an issue for the Passport website to have them reference the right version. 

https://github.com/passport/www.passportjs.org/issues/112

**Comment by rfossella** [comment]
srd90 - More background on the use of a username.   When a user enters our system we do not initially know what Idp credentials (company) they are associated with.  We are a multi-tenant system.  So user user1@xyz.com enters their name and presses enter;  we lookup user1@xyz.com and get the idp creds that are associated with xyz.com.  That's how the unique idp creds are gotten.

Each of those users are potentially associated with separate MS Entra accounts.  So when MS pops up their initial screen you choose user1@xyz.com.  User2 maybe user2@abc.com with completely different idp creds.   Again, mutli-tenant system w/different idp creds.

Note: this was tested 100's of times w/**passport-saml** package, with no cert errors not the latest "document signature" errors.  The only issue is what my original post stated, that  sometimes I would be holding onto an old profile.

That said, these new issues came up with **@node-saml/node-saml** installed.  You pointed me to the use of **idpcert** instead of **cert** for my MSEntra IDP.  FYI: Our users/companies may not (probably won't) **all** use MSEntra.  That certainly fixed the cert error when I login the first user.  

All my other logic is the same.  I trusted that the new version would be a little more seamless.

I hope I cleared up the use of username and our multi-tenant system.

I also saw their is a multistrategy option available.  I planned to look into that once i had the new version working like the old version - obviously this where my snag is.

I will take a look at some of your other suggestions in you post.

**Comment by srd90** [comment]
>The only issue is what my original post stated, that sometimes I would be holding onto an old profile.

passport-saml, @node-saml/passport-saml and @node-saml/node-saml are stateless. Problem is related to either to your lack of SLO (i.e. performing only local logout) and hitting to IdP with IdP's cookies indicating to IdP that UE/browser has already existing session with IdP which trigger passive authentication (i.e. IdP delivers - from your point of view - old user info). Not an issue of passport-saml or scoped passport-saml. Anorher source of problem might be session management solution that you have on top of authentication subsystem (passport-saml).

Eitherway the fact that you sometimes get incorrect autentication should have triggered you to put your system immediately offline (dunno though how sensitive information that system holds).

>You pointed me to the use of **idpcert** instead of **cert** for my MSEntra IDP. FYI: Our users/companies may not (probably won't) all use MSEntra. That certainly fixed the cert error when I login the first user.

I pointed you to use **idpCert** because **cert** parameter was renamed to **idpCert** starting from 5.0.0. Since you got an error that **idpCert** was missing you must had been upgraded directly to version where that rename was already introduced (it was 5.0.0)

* https://github.com/node-saml/node-saml/blob/v5.0.0/CHANGELOG.md#-major-changes
* https://github.com/node-saml/node-saml/pull/343

Suggestion was not specific/related only to MSEntra. It is the way forward with any IdP with versions >= 5.0.0

If your login worked with msentra and you had not set both "wants" to false then maybe that idp signs response and assertion and maybe your second user's IdP doesn't. That is one possibility to your document signature issue.

>All my other logic is the same. I trusted that the new version would be a little more seamless.

If you would read through changelog between 3.2.4 all the way through 4.x to 5.0.0 you would spot multiple major / breaking changes which introduce stricter default configuration. You have option to e.g. loosen whether response and/or assertion level signatures are required or whether it is suffient to have one of those regardless of which one (and thats how passport-saml < 4.0.0 worked...those versions did not care and certain old versions of passport-saml were possible to configure to skip signature validation completely which meant that anyone could easily impersonate anyone else by simply posting any saml response).

Just in case you wonder/wondered whether setting both "wants" to false would/should turn off signature validation (even though documentation states otherwise along with freely readable codebase) then here is answer to similar question: https://github.com/node-saml/node-saml/issues/326#issuecomment-1717221811

Just in case you do not know what to look for (at wire level) when you try to figure how IdP is configure see content of this comment: https://github.com/node-saml/passport-saml/discussions/671#discussioncomment-5261103 (I deliberetely did not list false&false so that copy paste coders would take a moment to think implications before doing so).

**Comment by rfossella** [comment]
I read through all the valuable comments.  The clients that use our application must be white-listed so even the first iteration, not requiring assertion level signatures, was satisfactory.

I did loosen up the restrictions using the options, and I'm no longer getting the signature error.  _Part_ of the issue with this is instructing our clients to make the correct settings in their respective Idp - not that it is necessarily a bad thing.
```
wantAssertionsSigned: false,
wantAuthnResponseSigned: false
```
So I at least have a stable working QA version using latest release.

FYI: the "old" profile issue remains, but as you stated, it certainly could be related to other things outside of passport-saml.  Technically speaking, this issue primarily comes up in my testing suite since I have about six different users that I test with that are in and out of "signed-in status."  Most, if not all of our users, use a single client login user id.

That said, I did add an _optional_ UI/UX option to the app that will let the user know there are multiple "signed-in users" trying to access same app w/same IdP credentials.  In this case I give them the option to SLO the _problematic_ user.  I discussed w/Product Management and also forcing a SLO when the user _only_ meant to sign out of local app is not what they want.  They are aware of the possible snag.  If I figure out anything more about that issue I will share in the forum.

Thank you again for all your help with this.  Very much appreciated.

**Comment by rfossella** [comment]
The discussion has been helpful and addressed all questions/concerns.

**Reply by ozgencDev** [reply]
Hi Mr Fosella Did you solve that problem, I'm experiencing exactly the same thing.

**Reply by srd90** [reply]
AFAIK discussion about this case continued here:
* https://github.com/node-saml/passport-saml/discussions/944

**Reply by ozgencDev** [reply]
Hi @srd90 , I read that discussion, but I didn’t find a solution for my particular case, similar to Mr. Fosella’s problem. The Profile Object retrieves an old user. How can I solve this problem? and also  I am using forceAuthn but azure login not come sometimes ? Azure retrieve wrong profile to me and I am also using relay state
```
getSamlOptions: function(req, done) {
                    let email = _.get(req,'query.RelayState') || _.get(req,'body.RelayState') || _.get(req,'user.email');
                    let domain = email.split('@')[1];
                    return customerCtrl.getCustomerSamlConfigByIdentifier(
                        domain
                    ).then(samlConfig => {
                        return done(null, {
                            //path: 'saml/login/callback',
                            callbackUrl: 'https://localhost:7788/saml/login/callback',
                            entryPoint: samlConfig.ssoUrl,
                            issuer: samlConfig.entityId,
                            idpCert: samlConfig.cert,
                            wantAuthnResponseSigned: false,
                            logoutCallbackUrl: 'https://localhost:7788/saml/logout/callback'
                            //callbackUrl:  process.env.SUSTAIN_BASE_URL + '/backend/saml/login/callback'
                        });
                    })
                    
                }
            },
 ```

**Reply by srd90** [reply, marked as answer]
>How can I solve this problem?

First of all `@node-saml/node-saml` and `@node-saml/passport-saml` are stateless. Your session middleware and browser keep some state in the form of cookies set by your application's session middleware and IdP's session subsystem to track SSO and SLO sessions.

I would suggest you to collect full trace (HTTP messages) of your problem scenario so that trace contains session(s) prior to and "failed login". Pay attention to session cookies transferred between IdP and your browser and pay attention to content of SAML messages contents (auth responses). Most probably your browser sends to IdP "old user's" SSO session cookie and IdP does what it is suppose to do i.e. performs noninteractive login (and returns "old users" data). Most probably you fail to logout from IdP (or if you do not perform SLO then your IdP works as it is expected to work). You are on your own on this because you must NOT share IdP auth responses in any form for outsiders (unless you spend time to redact those and if you decide to do so pay attention to compressed and base64 encoded contents also).

Quote from https://github.com/node-saml/node-saml/discussions/382#discussioncomment-11275691
>Use some SAML tracer plugin at your browser to see whats being transferred between IdP and SP.
>E.g. this https://chromewebstore.google.com/detail/saml-tracer/mpdajninpobndbfcldcmbpnnbhibjmch

About this:
>Azure retrieve wrong profile to me and I am also using relay state

FWIW, RelayState is just a string echoed back to SP (content is meaningfull only for your application layer). It is not involved to any session tracking from SAML protocol or IdP point of view.

About rfossella's case...rfossella did not want to do SLO when he wanted to change users and furthermore used same browser instance to perform logins to various IdPs. I.e. result (utter mess) is what he wanted and what he got.

If you are (trying to) performing SLO for users then I would suggest you to inspect your traffic trace to/from IdP and SP and ask yourself whether this could be the case of blocked 3rd party cookies or are you sending correct information to IdP if/when your SP triggers SLO (sends LogoutRequest to IdP after it had performed successfull local logout).

**Reply by ozgencDev** [reply]
Thanks for your reply, @srd90. I recently achieved a successful logout, and this solved half of my problem. However, if a user logs in with multiple accounts (for example, three different accounts), when I log out of the 'old user', there are still two active accounts, and the current account switches to one of them. I think force authn solves this problem with correct account selection, but force authn does not always come. No matter what, I would like to bring up the account selection screen and I guess that's what forceauthn is for  but as I said forceAuthn not trigger always

**Comment by rfossella** [comment]
Please note that in MY case (company policy) we _specifically_ **did not** want to do SLO.  Outside clients who set up their Azure do not include SLO (instructions/url), nor did we want it in-house.  We do the same thing for JIRA, Confluence, etc.  For cases when (our) app wants to _force_ an SLO, we implement a shortcut for that.  Not exactly sure what the _utter mess_ is, but @srd90 did put me on the right track with RelayState.  In this case I could verify that the _entered_ username matched IdP profile.
